neuralop/layers/attention_kernel_integral.py

```diff
@@ -0,0 +1,228 @@
+        self.init_weight(self.to_k, init_fn)
+        self.init_weight(self.to_v, init_fn)
+
+    def normalize_wrt_domain(self, u, norm_fn):
+        """
+        Normalize the input function with respect to the domain,
+         reshape the tensor to [batch_size*n_heads, num_grid_points, head_n_channels]
+        The second dimension is equal to the number of grid points that discretize the domain
+        """
+        # u: the input or transformed function
+        batch_size = u.shape[0]
+        u = u.view(batch_size*self.n_heads, -1, self.head_n_channels)
+        u = norm_fn(u)    # layer norm with channel dimension or instance norm with spatial dimension
+        return u.view(batch_size, self.n_heads, -1, self.head_n_channels)
+
+    def forward(self,
+                u_src,
+                pos_src,
+                positional_embedding_module=None,    # positional encoding module for encoding q/k
+                u_qry=None,
+                pos_qry=None,
+                weights=None,
+                associative=True,   # can be much faster if num_grid_points is larger than the channel number c
+                return_kernel=False):
+        """
+        Computes kernel integral transform with attention
+
+        Parameters
+        ----------
+        u_src: input function (used to compute key and value in attention),
+                tensor of shape [batch_size, num_grid_points_src, channels]
+        pos_src: coordinate of the second source of function's sampling points y,
+                tensor of shape [batch_size, num_grid_points_src, pos_dim]
+        positional_embedding_module: positional embedding module for encoding query/key (q/k),
+                a torch.nn.Module
+        u_qry: query function,
+                tensor of shape [batch_size, num_grid_points_query, channels], if not provided, u_qry = u_src
+        pos_qry: coordinate of query points x,
+                tensor of shape [batch_size, num_grid_points_query, pos_dim], if not provided, pos_qry = pos_src
+        weights : tensor of shape [batch_size, num_grid_points_src], if not provided assume to be 1/num_grid_points_src
+                Weights for each point y proportional to the volume around v(y) = u_src(y) W_v being integrated,
+                where W_v is the learnable linear projection for value in attention.
+        associative: if True, use associativity of matrix multiplication, first multiply K^T V, then multiply Q,
+                much faster when num_grid_points is larger than the channel number (which is usually the case)
+        return_kernel: if True, return the kernel matrix (for analyzing the kernel)
+
+        Output
+        ----------
+        u: Output function given on the query points x.
+        """
+
+        if u_qry is None:
+            u_qry = u_src   # go back to self attention
+        else:
+            assert pos_qry is not None, 'query coordinates are required if query function is provided'
```

Comment:
Can we raise a `ValueError` here and also for the `Exception` below?

Reply:
Yes